src/crypto/hsm/nxp/CHIPCryptoPALHsm_SE05X_P256.cpp

```diff
@@ -482,6 +495,216 @@ CHIP_ERROR P256PublicKeyHSM::ECDSA_validate_hash_signature(const uint8_t * hash,
+ *
+ *                (ASN1_CONSTRUCTED | ASN1_SET) LENGTH
+ *
+ *                    (ASN1_CONSTRUCTED | ASN1_SEQUENCE) LENGTH
+ *
+ *                        (ASN1_OID) LENGTH VALUE(Organisation OID)
+ *
+ *                        (ASN1_UTF8_STRING) LENGTH VALUE(Subject Str == "CSR")
+ *
+ *            PUBLIC KEY {WITH HEADER. 91 Bytes}
+ *
+ *        (ASN1_CONSTRUCTED | ASN1_SEQUENCE) LENGTH
+ *
+ *            (ASN1_OID) LENGTH VALUE(ECDSA SHA256 OID)
+ *
+ *            (ASN1_NULL) 0x00
+ *
+ *        (ASN1_BIT_STRING) LENGTH VALUE(SIGNATURE)
+ *
+ */
+
+CHIP_ERROR P256KeypairHSM::NewCertificateSigningRequest(uint8_t * csr, size_t & csr_length)
+{
+    CHIP_ERROR error           = CHIP_ERROR_INTERNAL;
+    sss_status_t status        = kStatus_SSS_Success;
+    sss_asymmetric_t asymm_ctx = { 0 };
+    sss_object_t keyObject     = { 0 };
+    sss_digest_t digest_ctx    = { 0 };
+
+    uint8_t data_to_hash[128] = { 0 };
+    size_t data_to_hash_len   = sizeof(data_to_hash);
+    uint8_t pubkey[128]       = { 0 };
+    size_t pubKeyLen          = 0;
+    uint8_t hash[32]          = { 0 };
+    size_t hash_length        = sizeof(hash);
+    uint8_t signature[128]    = { 0 };
+    size_t signature_len      = sizeof(signature);
+
+    size_t csr_index    = 0;
+    size_t buffer_index = data_to_hash_len;
+
+    uint8_t organisation_oid[3] = { 0x55, 0x04, 0x0a };
+
+    // Version  ::=  INTEGER  {  v1(0), v2(1), v3(2)  }
+    uint8_t version[3]       = { 0x02, 0x01, 0x00 };
+    uint8_t signature_oid[8] = { 0x2a, 0x86, 0x48, 0xce, 0x3d, 0x04, 0x03, 0x02 };
+    uint8_t nist256_header[] = { 0x30, 0x59, 0x30, 0x13, 0x06, 0x07, 0x2A, 0x86, 0x48, 0xCE, 0x3D, 0x02, 0x01,
+                                 0x06, 0x08, 0x2A, 0x86, 0x48, 0xCE, 0x3D, 0x03, 0x01, 0x07, 0x03, 0x42, 0x00 };
+
+    ChipLogDetail(Crypto, "NewCertificateSigningRequest: Using SE05X for creating CSR !");
+
+    // No extensions are copied
+    buffer_index = buffer_index - 2;
+    add_tlv(data_to_hash, buffer_index, (ASN1_CONSTRUCTED | ASN1_CONTEXT_SPECIFIC), 0, NULL);
+
+    // Copy public key (with header)
+    {
+        P256PublicKeyHSM & public_key = const_cast<P256PublicKeyHSM &>(Pubkey());
+
+        VerifyOrExit((sizeof(nist256_header) + public_key.Length()) <= sizeof(pubkey), error = CHIP_ERROR_INTERNAL);
+
+        memcpy(pubkey, nist256_header, sizeof(nist256_header));
+        pubKeyLen = pubKeyLen + sizeof(nist256_header);
+
+        memcpy((pubkey + pubKeyLen), Uint8::to_uchar(public_key), public_key.Length());
+        pubKeyLen = pubKeyLen + public_key.Length();
+    }
+
+    buffer_index = buffer_index - pubKeyLen;
+    VerifyOrExit(buffer_index > 0, error = CHIP_ERROR_INTERNAL);
+    memcpy((void *) &data_to_hash[buffer_index], pubkey, pubKeyLen);
+
+    // Copy subject (in the current implementation only organisation name info is added) and organisation OID
+    buffer_index = buffer_index - (1 + 1 + sizeof(SUBJECT_STR) - 1);
```

Comment:
I don't really like how the buffer_index is decrease/calculated through this function. It seems error prone.
The use of constants could help make this clearer 
buffer_index -= kTlvHeader

Reply:
@sujaygkulkarni-nxp ?

Reply:
Sure. Updated the code by using kTlvHeader,